echopop/live/live_biology.py

```diff
@@ -0,0 +1,1377 @@
+from functools import reduce
+
+import numpy as np
+import pandas as pd
+
+from ..utils.operations import group_interpolator_creator
+from .live_acoustics import average_sigma_bs
+from .live_spatial_methods import apply_spatial_definitions
+from .sql_methods import (
+    SQL,
+    get_table_key_names,
+    sql_data_exchange,
+    sql_group_update,
+    sql_update_strata_summary,
+)
+
+
+def biology_data_filter(biology_data: pd.DataFrame, filter_dict: dict):
+
+    # Create dataframe copy
+    data_copy = biology_data.copy()
+
+    # Iterate through dictionary to apply filters (if present)
+    for column, value in filter_dict.items():
+        if column in data_copy.columns:
+            data_copy = data_copy[data_copy[column] == value]
+
+    # Return output
+    return data_copy
+
+
+def merge_trawl_info(biology_dict: dict):
+
+    # Get the trawl information dictionary
+    trawl_info_df = biology_dict["trawl_info_df"]
+
+    # Update `catch_df`
+    biology_dict["catch_df"] = biology_dict["catch_df"].merge(trawl_info_df)
+
+    # Update `length_df`
+    biology_dict["length_df"] = biology_dict["length_df"].merge(trawl_info_df)
+
+    # Update `specimen_df`
+    biology_dict["specimen_df"] = biology_dict["specimen_df"].merge(trawl_info_df)
+
+    # Drop the trawl information
+    del biology_dict["trawl_info_df"]
+
+
+def prepare_length_distribution(file_configuration: dict):
+
+    # Get the length distribution parameters
+    distrib_params = file_configuration["biology"]["length_distribution"]["bins"]
+
+    # Create histogram bins
+    length_bins = np.linspace(
+        **{key: value for key, value in zip(["start", "stop", "num"], distrib_params)}, dtype=float
+    )
+
+    # Get the binwidths
+    binwidth = np.diff(length_bins / 2.0).mean()
+
+    # Generate the equivalent interval boundaries for each bin
+    intervals = np.concatenate([length_bins[:1] - binwidth, length_bins + binwidth])
+
+    # Format as a DataFrame and return the output
+    # ---- Add Categorical interval column
+    length_bins_df = pd.DataFrame(
+        {"length_bin": length_bins, "interval": pd.cut(length_bins, intervals)}
+    )
+    # ---- Add numeric lower boundary
+    length_bins_df["lower"] = length_bins_df["interval"].apply(lambda x: x.left).astype(float)
+    # ---- Add numeric upper boundary
+    length_bins_df["upper"] = length_bins_df["interval"].apply(lambda x: x.right).astype(float)
+
+    # Return the dataframe that will be incorporated into the biological data attribute
+    return length_bins_df
+
+
+def preprocess_biology_data(biology_output: dict, spatial_dict: dict, file_configuration: dict):
```

Comment:
I need your help to go through the intention of this function, because it is quite hidden behind the many layers of functions that are very specific. I think there's a balance between reusability and readability that we should discuss. I also have an additional question on why we need a database if for new trawls we are basically just appending (accumulating) new data.